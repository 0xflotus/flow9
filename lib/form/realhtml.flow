import form/form;
import sys/target;
import form/realhtmlworkaround;
export {
	//  A wrapper for RealHTML form with channels to send/receive cross domain messages with JS window.postMessage
	// function.
	//  targetOrigin :
	// Specifies what the origin of embedded page must be for the event to be dispatched, either as the literal string "*"
	// (indicating no preference) or as a URI. If at the time the event is scheduled to be dispatched the scheme,
	// hostname, or port of otherWindow's document does not match that provided in targetOrigin,
	// the event will not be dispatched

	//	An example of JS for embedded cross-domain page
/*
<script type="text/javascript" src="/flow/js/toflow.js"></script>
<script type="text/javascript">
	function receiveMessage(e) {
		// When sendChannel is changed in flow code, this is where the message goes:
		console.log("Message from panel received: " + e.data);
	}
	window.addEventListener('message', receiveMessage);
</script>
*/
	//	Sending messages to flow code:
	//		postToFlow("message to flow")
	// You can send any object which is converted to a JSON string and
	// received by the flow side through the receiveChannel.

	RealHTMLCrossDomain(url : string, domain : string, wh : Behaviour<WidthHeight>,
		sendChannel : DynamicBehaviour<string>,
		receiveChannel : DynamicBehaviour<string>,
		targetOrigin : string,
		key : string,
		onLoaded : () -> void,
		zoomEnabled : Behaviour<bool>) -> Form;

	// + onError, which is not very reliable, see form.flow, OnError style for RealHtml
	RealHTMLCrossDomain2(url : string, domain : string, wh : Behaviour<WidthHeight>,
		sendChannel : DynamicBehaviour<string>,
		receiveChannel : DynamicBehaviour<string>,
		targetOrigin : string,
		key : string,
		onLoaded : () -> void,
		onError : (string) -> void,
		zoomEnabled : Behaviour<bool>) -> Form;

	isCustomCacheSupported() -> bool;
}

RealHTMLCrossDomain(url, domain, wh, sendChannel, receiveChannel, targetOrigin, key, onLoaded, zoomEnabled) {
	RealHTMLCrossDomain2(url, domain, wh, sendChannel, receiveChannel, targetOrigin, key, onLoaded, ignore, zoomEnabled);
}

RealHTMLCrossDomain2(url : string, domain : string, wh : Behaviour<WidthHeight>, sendChannel : DynamicBehaviour<string>,
	receiveChannel : DynamicBehaviour<string>, targetOrigin : string, key : string, onLoaded : () -> void, onError : (string) -> void, zoomEnabled : Behaviour<bool>) -> Form {
	sendToJS = ref \m : string -> {};
	getHostCall = \hc -> {
		sendToJS := \msg : string -> {
			hc("postMessage", [msg, targetOrigin]);
			{}
		};
	};
	flowCallBack = \args -> {
		if (length(args) == 2 && args[0] == "postMessage") {
			next(receiveChannel, args[1]);
		}
		""
	};
	Constructor(
		makeRealHTML(key,
			RealHTML2(url, wh, [FlowCallback(flowCallBack), PageHostcallSetter(getHostCall),
				OverridePageDomain(domain), UpdateCachedContent(true), OnPageLoaded(onLoaded), OnError(onError), ZoomEnabled(zoomEnabled)])
		),
		\ -> { subscribe(sendChannel, \m -> ^sendToJS(m)) }
	);
}

isCustomCacheSupported() {
	ios(); // Is supported only for iOS for now.
}
