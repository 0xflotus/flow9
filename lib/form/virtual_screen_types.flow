import math/geometry;
import form/form;
import fusion;

export {
	VirtualScreenInfo : (
		size : Behaviour<WidthHeight>,
		position : Behaviour<PositionScale>
	);

	CropPopupByVirtualScreenBox(
		info : VirtualScreenInfo,
		popupContent : Form,
		popupSize : Behaviour<WidthHeight>,
		translateToVisibleArea : bool
	) -> Form;
}

CropPopupByVirtualScreenBox(
	info : VirtualScreenInfo,
	popupContent : Form,
	popupSize : Behaviour<WidthHeight>,
	translateToVisibleArea : bool
) -> Form {
	screenXB = make(0.0);
	screenYB = make(0.0);
	screenWB = make(0.0);
	screenHB = make(0.0);
	windowXB = make(0.0);
	windowYB = make(0.0);
	Constructor(
		popupContent
		|> (\f -> if (translateToVisibleArea) Translate(windowXB, windowYB, f) else f)
		|> (\f -> Crop(screenXB, screenYB, screenWB, screenHB, f))
		|> (\f -> Translate(screenXB, screenYB, f)),
		make3Subscribe(info.size, info.position, popupSize, \size, p, windowWH -> {
			nextDistinct(screenXB, p.pos.x);
			nextDistinct(screenYB, p.pos.y);

			screenWidth = size.width * p.scale.x;
			screenHeight = size.height * p.scale.y;

			nextDistinct(screenWB, screenWidth);
			nextDistinct(screenHB, screenHeight);
			nextDistinct(windowXB, p.pos.x + max(0.0, (screenWidth - windowWH.width) / 2.0));
			nextDistinct(windowYB, p.pos.y + max(0.0, (screenHeight - windowWH.height) / 2.0));
		})
	);
}
