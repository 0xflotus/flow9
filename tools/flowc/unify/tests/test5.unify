import form;

α14947 c= () -> α15363;
(α14947) -> α15365 c= (α14949) -> α14950;
() -> Empty c= α14949;
() -> Graphics c= α14949;

finalize
