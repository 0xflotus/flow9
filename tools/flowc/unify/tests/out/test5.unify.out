α101={<union> ::= (Empty<>, Graphics<>) .. *}
α14947=() -> α100
α14949=() -> {<union> ::= (Empty<>, Graphics<>) .. *}

Final:
α100=Form
α101=Form
α14947=() -> Form
α14949=() -> Form
α14950=?
α15363=Form
α15365=?
